src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateBuilder.java

```diff
@@ -311,12 +314,27 @@ public Pod build() {
 
             KubernetesCloud cloud = slave.getKubernetesCloud();
 
-            String url = cloud.getJenkinsUrlOrDie();
-
-            env.put("JENKINS_URL", url);
             if (!StringUtils.isBlank(cloud.getJenkinsTunnel())) {
                 env.put("JENKINS_TUNNEL", cloud.getJenkinsTunnel());
             }
+            
+            if (!cloud.isDirectConnection()) {
+                env.put("JENKINS_URL", cloud.getJenkinsUrlOrDie());
+            } else {
+                Jenkins jenkins = Jenkins.getInstanceOrNull();
+                if(jenkins == null) throw new IllegalStateException("Jenkins instance null");
+                String host;
+                try {
+                    host = new URL(jenkins.getRootUrl()).getHost();
```

Comment:
This should prefer `CLI_HOST_NAME` I think.

Reply:
Both `hudson.TcpSlaveAgentListener.CLI_HOST_NAME` and `SystemProperties.getString(TcpSlaveAgentListener.class.getName()+".hostName")`, which is used there, are marked with `@Restricted(NoExternalUse.class)`. I'm not sure if I can bypass this somehow.

Shall I keep the current implementation? Do you see any chances for wrong behaviour?

Reply:
You could look up the system property, which would work assuming there is nothing overriding this field using a Groovy init script or the like. Pair with a PR to `jenkinsci/jenkins` adding a `getAdvertisedHost` method, and a TODO comment here reminding us to use the official API once the Jenkins baseline is sufficiently new.

Reply:
I created https://github.com/jenkinsci/jenkins/pull/4227 and adjusted the code accordingly.

Reply:
https://github.com/jenkinsci/jenkins/pull/4227 is merged and [released](https://github.com/jenkinsci/jenkins/releases/tag/jenkins-2.198).
But I would prefer not to consume the change yet since we do not plan to upgrade our Jenkins to this version. I would propose to wait until the change is available in a new Jenkins LTS version - this would be OK for us.